File: Fall_2026/demos/03_collections_demo.py
"""
Session 3 - Collections and Indexing. Python side.

Run this ALONGSIDE demos/03_collections_demo.R, one section at a time, with
both consoles visible. The comparison IS the lesson.

    python demos/03_collections_demo.py
"""


def rule(title):
    print("\n" + "=" * 62)
    print(title)
    print("=" * 62)


# ------------------------------------------------- slide 5: build a list
rule("Lists")

bp = [117, 122, 141, 130, 118]
print(bp)
print("length:", len(bp))


# ------------------------------------ slide 6: Python refuses to coerce
rule("Python does NOT silently coerce")

mixed = [1, 2, "three"]
print(mixed, "-> stays mixed, each element keeps its own type")
print("types:", [type(x).__name__ for x in mixed])
print("\nCompare with R, which turns the whole thing into text without a word.")


# ------------------------------------------------ slides 7-9: indexing
rule("Indexing: Python counts from 0")

print("bp[0]  ->", bp[0], "  (first)")
print("bp[1]  ->", bp[1], "  (second)")
print("bp[-1] ->", bp[-1], "  <- the LAST one. In R this drops the first.")

print("\nbp[1:4] ->", bp[1:4], " (stop is EXCLUSIVE - indices 1,2,3)")
print("bp[:3]  ->", bp[:3], " (first three)")
print("bp[-2:] ->", bp[-2:], "      (last two)")
print("bp[:0]  ->", bp[:0], "           (empty - not an error!)")

print("\nWhy exclusive? So these two split the list with no overlap and no gap:")
print("  bp[:3] + bp[3:] ==", bp[:3] + bp[3:])
print("  and len(bp[1:4]) == 4 - 1 ==", len(bp[1:4]))


# ------------------------------------- slide 11: masks need numpy/pandas
rule("Boolean indexing needs numpy in Python")

print("Plain lists do not vectorise:")
try:
    print(bp >= 130)
except TypeError as e:
    print("  bp >= 130  -> TypeError:", e)
print("\n--> THIS is why numpy and pandas exist, and why Session 7 starts there.")

try:
    import numpy as np
except ImportError:
    print("\n(numpy not installed - skipping the vectorised half)")
else:
    arr = np.array(bp)
    print("\nWith numpy:")
    print("  arr >= 130       ->", arr >= 130)
    print("  arr[arr >= 130]  ->", arr[arr >= 130])
    print("  (arr >= 130).sum() ->", (arr >= 130).sum())
    print("\nIn R this works with no imports at all. Same idea, different packaging.")


# ------------------------------------------------------ slide 12: dicts
rule("Dicts - lookup by name, not position")

patient = {"id": "P042", "age": 61, "systolic": 141, "treated": True}

print('patient["age"]  ->', patient["age"])
patient["age"] = 62
print("after update    ->", patient["age"])
patient["dx"] = "HTN"
print("keys            ->", list(patient.keys()))

print("\nA dataframe row is essentially one of these. See you in Session 7.")

File: Fall_2026/demos/03_collections_demo.R
# Session 3 - Collections and Indexing. R side.
#
# Run this ALONGSIDE demos/03_collections_demo.py, one section at a time,
# with both consoles visible. The whole point of this session is the
# side-by-side comparison, so do not run them separately.
#
#   source("demos/03_collections_demo.R")

rule <- function(title) {
  cat("\n", strrep("=", 62), "\n", title, "\n", strrep("=", 62), "\n", sep = "")
}

# ------------------------------------------------- slide 5: build a vector
rule("Vectors: c() means combine")

bp <- c(117, 122, 141, 130, 118)
print(bp)
cat("length:", length(bp), "\n")

# ---------------------------------------------- slide 6: silent coercion
rule("R coerces silently - no warning, no error")

mixed <- c(1, 2, "three")
print(mixed)
cat("class:", class(mixed), "  <- everything became text\n")
cat("\nThis is what happens when one row of your age column says 'unknown'.\n")

# ------------------------------------------------ slides 7-9: indexing
rule("Indexing: R counts from 1")

cat("bp[1]  ->", bp[1], "   (first)\n")
cat("bp[2]  ->", bp[2], "   (second)\n")
cat("bp[-1] -> ", paste(bp[-1], collapse = " "), "\n", sep = "")
cat("          ^ DROPS the first element. In Python this means 'the last one'.\n")

cat("\nbp[2:4] ->", paste(bp[2:4], collapse = " "), "  (stop is INCLUSIVE)\n")
cat("bp[c(1,3,5)] ->", paste(bp[c(1, 3, 5)], collapse = " "), "  (cherry-pick)\n")
cat("bp[0]  -> ", paste(bp[0], collapse = " "), "(empty - not an error!)\n", sep = "")

# ------------------------------------------ slide 11: the boolean mask
rule("Boolean indexing - the idiom that runs the whole semester")

cat("Step 1, the mask:      ")
print(bp > 130)

cat("Step 2, use the mask:  ")
print(bp[bp > 130])

cat("Step 3, count it:      ")
print(sum(bp > 130))

cat("\nRead bp[bp > 130] as: 'the elements of bp, where bp is over 130'.\n")
cat("dplyr::filter() is this. df[df.age > 65] is this. Learn it here.\n")

# ------------------------------------------------ slide 12: named lists
rule("Named lists - lookup by name, not position")

patient <- list(id = "P042", age = 61, systolic = 141, treated = TRUE)

cat("patient$age        ->", patient$age, "\n")
patient$age <- 62
cat("after update       ->", patient$age, "\n")
patient$dx <- "HTN"
cat("names(patient)     ->", paste(names(patient), collapse = " "), "\n")
